libpod/image/image.go

```
@@ -169,15 +170,169 @@ func (ir *Runtime) New(ctx context.Context, name, signaturePolicyPath, authfile
+	if format != DockerArchive {
+		return errors.Errorf("multi-image archives are only supported in in the %q format", DockerArchive)
+	}
+
+	sys := GetSystemContext("", "", false)
+	multiDest, err := archive.NewMultiImageDestination(sys, outputFile)
+	if err != nil {
+		return err
+	}
+
+	policyContext, err := getPolicyContext(sys)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := policyContext.Destroy(); err != nil {
+			logrus.Errorf("failed to destroy policy context: %q", err)
+		}
+		if err := multiDest.Finalize(ctx); err != nil {
+			if finalErr == nil {
+				finalErr = err
+			} else {
+				finalErr = errors.Wrap(finalErr, err.Error())
+			}
+		}
+	}()
+
+	manifestMIMEType := multiDest.SupportedManifestMIMETypes()[0]
+
+	// Check whether c/image's progress bars use stderr or stdout.  Use
+	// stderr in case we need to be quiet or if the output is set to
+	// stdout.  If the output is set of stout, any log message there would
+	// corrupt the tarfile.
+	writer := os.Stdout
+	if quiet || outputFile == os.Stdout.Name() {
+		writer = os.Stderr
+	}
+
+	// Look up the images in the local storage and determine the tags for
+	// them.
+	imageTags := make(map[*Image][]string)
+	imageOrder := []*Image{}
+	for _, nameOrID := range namesOrIDs {
+		// Look up the name or ID in the local image storage.
+		localImage, err := ir.NewFromLocal(nameOrID)
+		if err != nil {
+			return err
+		}
+
+		tags, exists := imageTags[localImage]
+		if !exists {
+			// Preserve the relative order AND make the archives
+			// deterministic - iterating over go maps yields a
+			// random order.
+			imageOrder = append(imageOrder, localImage)
+		}
+
+		// Unless we referred to an ID, add the input as a tag.
+		id := localImage.ID()
+		if strings.HasPrefix(id, nameOrID) {
+			continue
+		}
+		tag, err := NormalizedTag(nameOrID)
+		if err != nil {
+			return err
+		}
+		tags = append(tags, tag.String())
```

`tag` already is a `reference.Named`, only for this to lose that and use a string representation, and `GetAdditionalTags` later to undo that again (and silently skip non-tag values it seems).